sofar/sofar.py

```diff
@@ -0,0 +1,2125 @@
+            files
+        ``'name'``
+            Return a list of the convention names without version
+        ``'name_version'``
+            Return a list of tuples containing the convention name and version.
+
+        The default is ``None`` which does not return anything
+
+    Returns
+    -------
+    See parameter `return_type`.
+    """
+    # directory containing the SOFA conventions
+    directory = os.path.join(os.path.dirname(__file__), "conventions")
+
+    # SOFA convention files
+    paths = [file for file in glob.glob(os.path.join(directory, "*.json"))]
+
+    if verbose:
+        print("Available SOFA conventions:")
+
+    conventions = []
+    versions = []
+    for path in paths:
+        fileparts = os.path.basename(path).split(sep="_")
+        conventions += [fileparts[0]]
+        versions += [fileparts[1][:-5]]
+        if verbose:
+            print(f"{conventions[-1]} (Version {versions[-1]})")
+
+    if return_type is None:
+        return
+    elif return_type == "path":
+        return paths
+    elif return_type == "name":
+        return conventions
+    elif return_type == "name_version":
+        return [(n, v) for n, v in zip(conventions, versions)]
+    else:
+        raise ValueError(f"return_type {return_type} is invalid")
+
+
+def read_sofa(filename, verify=True, version="latest"):
+    """
+    Read SOFA file from disk and convert it to SOFA object.
+
+    Numeric data is returned as floats or numpy float arrays unless they have
+    missing data, in which case they are returned as numpy masked arrays.
+
+    Parameters
+    ----------
+    filename : str
+        The filename. '.sofa' is appended to the filename, if it is not
+        explicitly given.
+    verify : bool, optional
+        Verify and update the SOFA object by calling :py:func:`~Sofa.verify`.
+        This helps to find potential errors in the default values and is thus
+        recommended. If reading a file does not work, try to call `Sofa` with
+        ``verify=False``. The default is ``True``.
+    version : str, optional
+        The version to which the API is updated.
+
+        ``'latest'``
+            Use the latest API and upgrade the SOFA file if required.
+        ``'match'``
+            Match the version of the sofa file.
+        str
+            Version string, e.g., ``'1.0'``. Note that this might downgrade
+            the SOFA object
+
+        The default is ``'latest'``
+
+    Returns
+    -------
+    sofa : Sofa
+        The SOFA object filled with the default values of the convention.
+    """
+
+    # check the filename
+    if not filename.lower().endswith('.sofa'):
+        filename += ".sofa"
+    if not os.path.isfile(filename):
+        raise ValueError(f"{filename} does not exist")
+
+    # attributes that are skipped
+    skip = ["_Encoding"]
+
+    # init list of custom attributes
+    custom = []
+
+    # open new NETCDF4 file for reading
+    with Dataset(filename, "r+", format="NETCDF4") as file:
```

Comment:
Unfortunately yes: otherwise the line
`file[var]._Encoding = "ascii"`
below would not work, which is the way to read string data suggested in the official documentation of the netCDF Python API